src/providers.rs

```diff
@@ -46,361 +39,219 @@ pub const LLAMA_ARBITRUM_ONE_HOSTNAME: &str = "arbitrum.llamarpc.com";
-pub const ALCHEMY_ETH_MAINNET_CREDENTIAL: &str = "/v2/zBxaSBUMfuH8XnA-uLIWeXfCx1T8ItkM";
-pub const ALCHEMY_ETH_SEPOLIA_CREDENTIAL: &str = "/v2/Mbow19DWsfPXiTpdgvRu4HQq63iYycU-";
-pub const ALCHEMY_ARBITRUM_ONE_CREDENTIAL: &str = "/v2";
-pub const ALCHEMY_BASE_MAINNET_CREDENTIAL: &str = "/v2";
-pub const ALCHEMY_OPTIMISM_MAINNET_CREDENTIAL: &str = "/v2";
-pub const BLOCKPI_ETH_MAINNET_CREDENTIAL: &str = "/v1/rpc/0edc81e20be23ddff051f61a97bb457ec7284a58";
-pub const BLOCKPI_ETH_SEPOLIA_CREDENTIAL: &str = "/v1/rpc/1fe987fddded17db50862311720ff444991d4dab";
-pub const BLOCKPI_ARBITRUM_ONE_CREDENTIAL: &str =
-    "/v1/rpc/a8b89a41d2a341e32ee7aefcb20820a7cbb65f35";
-pub const BLOCKPI_BASE_MAINNET_CREDENTIAL: &str =
-    "/v1/rpc/bd458bf9f28ed45c77823814a937c812d2efd260";
-pub const BLOCKPI_OPTIMISM_MAINNET_CREDENTIAL: &str =
-    "/v1/rpc/d54bfe59299d56b0cbb8b3c69bd122f4ab5ac654";
-
-pub fn get_default_providers() -> Vec<RegisterProviderArgs> {
-    vec![
-        RegisterProviderArgs {
+thread_local! {
+    pub static PROVIDERS: Vec<Provider> = vec![
+        Provider {
+            provider_id: 0,
             chain_id: ETH_MAINNET_CHAIN_ID,
-            hostname: CLOUDFLARE_HOSTNAME.to_string(),
-            credential_path: "/v1/mainnet".to_string(),
-            credential_headers: None,
-            cycles_per_call: 0,
-            cycles_per_message_byte: 0,
+            url_pattern: format!("https://{CLOUDFLARE_HOSTNAME}/v1/mainnet"),
+            header_patterns: vec![],
         },
-        RegisterProviderArgs {
+        Provider {
+            provider_id: 1,
             chain_id: ETH_MAINNET_CHAIN_ID,
-            hostname: ANKR_HOSTNAME.to_string(),
-            credential_path: "/eth".to_string(),
-            credential_headers: None,
-            cycles_per_call: 0,
-            cycles_per_message_byte: 0,
+            url_pattern: format!("https://{ANKR_HOSTNAME}/eth"),
+            header_patterns: vec![],
         },
-        RegisterProviderArgs {
+        Provider {
+            provider_id: 2,
             chain_id: ETH_MAINNET_CHAIN_ID,
-            hostname: PUBLICNODE_ETH_MAINNET_HOSTNAME.to_string(),
-            credential_path: "".to_string(),
-            credential_headers: None,
-            cycles_per_call: 0,
-            cycles_per_message_byte: 0,
+            url_pattern: format!("https://{PUBLICNODE_ETH_MAINNET_HOSTNAME}"),
+            header_patterns: vec![],
         },
-        RegisterProviderArgs {
+        Provider {
+            provider_id: 3,
             chain_id: ETH_MAINNET_CHAIN_ID,
-            hostname: BLOCKPI_ETH_MAINNET_HOSTNAME.to_string(),
-            credential_path: BLOCKPI_ETH_MAINNET_CREDENTIAL.to_string(),
-            credential_headers: None,
-            cycles_per_call: 0,
-            cycles_per_message_byte: 0,
+            url_pattern: format!("https://{BLOCKPI_ETH_MAINNET_HOSTNAME}/v1/rpc/{{API_KEY}}"),
```

Comment:
In order to run the test suite and continue using the EVM RPC canister in the local replica, we may want to include something along the lines of `default_url` and `default_headers` for each provider in case an API key is missing. Alternatively, we could require developers to always provide their own API keys, but this would create some additional friction for new devs.

Thoughts on how we could address this? Having a default URL and headers has the benefit of covering the edge case where someone tries to use an RPC provider before we can insert the API keys. On the other hand, it would add more complexity on the configuration side.

Reply:
I'm not sure I follow the question, why not keep using the initial defaults we already have?
```rust
pub const BLOCKPI_ETH_MAINNET_CREDENTIAL: &str = "/v1/rpc/0edc81e20be23ddff051f61a97bb457ec7284a58";
pub const BLOCKPI_ETH_SEPOLIA_CREDENTIAL: &str = "/v1/rpc/1fe987fddded17db50862311720ff444991d4dab";
pub const BLOCKPI_ARBITRUM_ONE_CREDENTIAL: &str =
    "/v1/rpc/a8b89a41d2a341e32ee7aefcb20820a7cbb65f35";
pub const BLOCKPI_BASE_MAINNET_CREDENTIAL: &str =
    "/v1/rpc/bd458bf9f28ed45c77823814a937c812d2efd260";
pub const BLOCKPI_OPTIMISM_MAINNET_CREDENTIAL: &str =
    "/v1/rpc/d54bfe59299d56b0cbb8b3c69bd122f4ab5ac654";
```

Reply:
We can do this for some of the providers with previously existing local API keys. The caveat is that we would now have to provision local replica API keys for Cloudflare, Ankr, and LlamaNodes. It seems like we could use a free tier for all of the default API keys so far, but this might not be possible for all RPC providers in the future.

This approach would imply adding a field such as `default_api_key` or a separate mapping for these default API keys. Do either of these sound good to you?

Reply:
Thanks Ryan for the explanation! I like your idea of `default_url` and I think we might be able to resolve this problem and the one from your [comment](https://github.com/internet-computer-protocol/evm-rpc-canister/pull/252#discussion_r1725762958) below regarding cycles in tests as follows:

1. Each provider has in addition an optional `public_url` (slight renaming from your suggestion)
2. We introduce a new intialization flag, e.g., `demo` (fee free to suggest other names). If set, then the `puclic_url` are used (error if a requested provider doesn't have a public url) and no cycles accounting is made.
WDYT?

Reply:
Sure! `demo` seems like a good name for clarifying the broader purpose of the flag. I also prefer `public_url` over `default_url` now that you mention it.